src/nxt_otel.c

```diff
@@ -0,0 +1,403 @@
+     * in cases where trace fails to generate or is missing
+     */
+}
+
+
+void
+nxt_otel_trace_and_span_init(nxt_task_t *task, nxt_http_request_t *r)
+{
+    r->otel->trace =
+        nxt_otel_rs_get_or_create_trace(r->otel->trace_id);
+    if (r->otel->trace == NULL) {
+        nxt_log(task, NXT_LOG_ERR, "error generating otel span");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_HEADER_STATE);
+}
+
+
+void
+nxt_otel_test_and_call_state(nxt_task_t *task, nxt_http_request_t *r)
+{
+    if (r == NULL || r->otel == NULL) {
+        return;
+    }
+
+    switch (r->otel->status) {
+    case NXT_OTEL_UNINIT_STATE:
+      return;
+    case NXT_OTEL_INIT_STATE:
+        nxt_otel_trace_and_span_init(task, r);
+        break;
+    case NXT_OTEL_HEADER_STATE:
+        nxt_otel_span_add_headers(task, r);
+        break;
+    case NXT_OTEL_BODY_STATE:
+        nxt_otel_span_add_body(r);
+        break;
+    case NXT_OTEL_COLLECT_STATE:
+      nxt_otel_span_collect(task, r);
+        break;
+    case NXT_OTEL_ERROR_STATE:
+        nxt_otel_error(task, r);
+        break;
+    }
+}
+
+
+// called in nxt_http_request_error
+void
+nxt_otel_request_error_path(nxt_task_t *task, nxt_http_request_t *r) {
+  if (r->otel->trace == NULL) {
+        return;
+    }
+
+    // response headers have been cleared
+    nxt_otel_propagate_header(task, r);
+
+    // collect span immediately
+    if (r->otel) {
+        nxt_otel_state_transition(r->otel, NXT_OTEL_COLLECT_STATE);
+    }
+    nxt_otel_test_and_call_state(task, r);
+}
+
+
+nxt_int_t
+nxt_otel_parse_traceparent(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    char               *copy;
+    nxt_http_request_t *r;
+
+    /* For information on parsing the traceparent header:
+     * https://www.w3.org/TR/trace-context/#traceparent-header
+     * A summary of the traceparent header value format follows:
+     * Traceparent: "$a-$b-$c-$d"
+     *   a. version (2 hex digits) (ff is forbidden)
+     *   b. trace_id (32 hex digits) (all zeroes forbidden)
+     *   c. parent_id (16 hex digits) (all zeroes forbidden)
+     *   d. flags (2 hex digits)
+     */
+
+    r = ctx;
+    if (field->value_length != NXT_OTEL_TRACEPARENT_LEN) {
+        goto error_state;
+    }
+
+    /* strsep is destructive so we make a copy of the field
+     */
+    copy = nxt_mp_zalloc(r->mem_pool, field->value_length + 1);
+    if (copy == NULL) {
+        goto error_state;
+    }
+    memcpy(copy, field->value, field->value_length);
+
+    r->otel->version = (u_char *) strsep(&copy, "-");
+    r->otel->trace_id = (u_char *) strsep(&copy, "-");
+    r->otel->parent_id = (u_char *) strsep(&copy, "-");
+    r->otel->trace_flags = (u_char *) strsep(&copy, "-");
+
+    if (r->otel->version     == NULL ||
+        r->otel->trace_id    == NULL ||
+        r->otel->parent_id   == NULL ||
+        r->otel->trace_flags == NULL)
```

Comment:
While I certainly prefer breaking the line after the operator, In Unit we do it the other and start with the operator on a new line.